File: javascript/TestUtil.js
/**
 * @license
 * Copyright 2017 Google Inc. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Some utilities for all Javascript tests.
 */
goog.provide('wycheproof.TestUtil');
goog.require('goog.crypt');
goog.require('goog.crypt.base64');
goog.require('goog.json');

/**
 * Reads test vectors in JSON format from a file.
 * @param {!string} filename
 *
 * @return {!JSONObject}
 */
wycheproof.TestUtil.readJsonTestVectorsFromFile = function(filename){
  var fileContent = goog.loadFileSync_(filename);
  assertTrue('Invalid file format (expected JSON)', goog.json.isValid(fileContent));
  return JSON.parse(fileContent);
};


/**
 * Analyzes the result from a goog.testing.TestCase execution.
 *
 * @param {!goog.testing.TestCase.Result} result
 */
wycheproof.TestUtil.checkTestCaseResult = function(result) {
  var failMsg = '';
  if (result.errors.length > 0) {
    for (var i = 0; i < result.errors.length; i++) {
      failMsg += result.errors[i].message + '\n';
    }
    fail(failMsg);
  }
};


/**
 * Checks whether the given string is in hex format.
 * @param {!string} s
 *
 * @return {!boolean}
 */
wycheproof.TestUtil.isHex = function(s) {
  return /(^[0-9A-F]*$)|(^[0-9a-f]*$)/.test(s);
};

/**
 * Converts a hex string to a ArrayBuffer.
 * @param {!string} s
 *
 * @return {!ArrayBuffer}
 */
wycheproof.TestUtil.hexToArrayBuffer = function(s) {
  var bytes = goog.crypt.hexToByteArray(s);
  return new Uint8Array(bytes).buffer;
};

/**
 * Converts an ArrayBuffer to hex string
 * @param {!ArrayBuffer} ab
 *
 * @return {!string}
 */
wycheproof.TestUtil.arrayBufferToHex = function(ab) {
  return goog.crypt.byteArrayToHex(new Uint8Array(ab));
};

/**
 * Converts a Base64URL string to hex format
 * @param {!string} s
 *
 * @return {!string}
 */
wycheproof.TestUtil.base64UrlToHex = function (s) {
  var b64str = wycheproof.TestUtil.base64UrlToBase64(s);
  return wycheproof.TestUtil.base64ToHex(b64str);
};

/**
 * Converts a hex string to Base64 format.
 * @param {!string} s
 *
 * @return {!string}
 */
wycheproof.TestUtil.hexToBase64 = function(s) {
  var bytes = goog.crypt.hexToByteArray(s);
  return goog.crypt.base64.encodeByteArray(bytes);
};

/**
 * Converts a base64 string to hex format.
 * @param {!string} s
 *
 * @return {!string}
 */
wycheproof.TestUtil.base64ToHex = function(s) {
  var bytes = goog.crypt.base64.decodeStringToByteArray(s);
  return goog.crypt.byteArrayToHex(bytes);
};

/**
 * Converts a Base64 string to Base64URL
 * @param {!string} s
 *
 * @return {!string}
 */
wycheproof.TestUtil.base64ToBase64Url = function(s) {
  return s.replace(/\+/g, '-')
      .replace(/\//g, '_')
      .replace(/=/g, '');
};

/**
 * Converts a Base64Url string to Base64
 * @param {!string} s
 *
 * @return {!string}
 */
wycheproof.TestUtil.base64UrlToBase64 = function(s) {
  return (s + '==='.slice((s.length + 3) % 4))
      .replace(/\-/g, '+')
      .replace(/_/g, '/');
};

File: src/test_util.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Test util for cryptography.hazmat"""

__author__ = "bleichen@google.com (Daniel Bleichenbacher)"

import json
import re
import tarfile
from typing import Optional

from google3.pyglib import resources
from cryptography.hazmat.primitives import hashes

TEST_VECTOR_PATH = "google3/third_party/wycheproof/testvectors/"
TAR_PATH = "google3/experimental/users/bleichen/wycheproof/py3/hazmat/"

def get_test_vectors(fname: str,
                     path: Optional[str] = None):
  if path is None:
    path = TEST_VECTOR_PATH
  txt = resources.GetResource(path + fname)
  return json.loads(txt)


def get_all_test_vectors(pattern: str,
                         schema: Optional[str] = None,
                         path: Optional[str] = None):
  if path is None:
    path = TEST_VECTOR_PATH
  regex = re.compile(pattern)
  for root, dirs, files in resources.WalkResources(path):
    if root != path:
      # skip
      continue
    for name in files:
      if regex.match(name):
        try:
          test = get_test_vectors(name, path=path)
          if schema is None or test["schema"] == schema:
            yield name, test
        except Exception as e:
          print("Could not load %s. Reason: %s" % (name, e))

def extract(tar, name: str) -> bytes:
  f = tar.extractfile(name)
  txt = f.read()
  f.close()
  return txt

def extract_from_tar(pattern: str,
                     filename: str,
                     path: Optional[str] = None):
  """Returns the content of all members matching pattern in a tar file.
  
  Args:
    pattern: a regex pattern
    filename: the name of a tar file
  Yields:
    tuples (name:str, content: bytes)
  """
  if path is None:
    path = TAR_PATH
  file_obj = resources.GetResourceAsFile(path + filename)
  tar = tarfile.open(fileobj=file_obj)
  regex = re.compile(pattern)
  for member in tar:
    if regex.match(member.name):
      pem = extract(tar, member.name)
      yield member.name, pem


def get_hash(md):
  try:
    if md == "MD5":
      return hashes.MD5()
    elif md == "SHA-1":
      return hashes.SHA1()
    elif md == "SHA-224":
      return hashes.SHA224()
    elif md == "SHA-256":
      return hashes.SHA256()
    elif md == "SHA-384":
      return hashes.SHA384()
    elif md == "SHA-512":
      return hashes.SHA512()
    elif md == "SHA3-224":
      return hashes.SHA3_224()
    elif md == "SHA3-256":
      return hashes.SHA3_256()
    elif md == "SHA3-384":
      return hashes.SHA3_384()
    elif md == "SHA3-512":
      return hashes.SHA3_512()
    # Added in version 2.5, but it is always possible that
    # the backend does not support the hash.
    elif md == "SHA-512/224":
      return hashes.SHA512_224()
    elif md == "SHA-512/256":
      return hashes.SHA512_256()
  except AttributeError:
    raise ValueError("Unsupported hash:" + md)
  raise ValueError("unknown hash " + md)
